Microsoft.Identity.Web/Client/TokenCacheProviders/Redis/MSALPerUserRedisTokenCacheProvider.cs

```diff
@@ -0,0 +1,135 @@
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Identity.Client;
+using System.Security.Claims;
+using System;
+
+namespace Microsoft.Identity.Web.Client.TokenCacheProviders
+{
+    /// <summary>
+    /// An implementation of token cache for both Confidential and Public clients backed by Redis Cache.
+    /// When cached data is distributed, the data:
+    /// Is coherent(consistent) across requests to multiple servers.
+    /// Survives server restarts and app deployments.
+    /// Doesn't use local memory.
+    /// </summary>
+    /// <seealso cref="https://aka.ms/msal-net-token-cache-serialization"/>
+    public class MSALPerUserRedisTokenCacheProvider : IMSALUserTokenCacheProvider
+    {
+        /// <summary>
+        /// The backing DistributedCache instance
+        /// </summary>
+        internal IDistributedCache distributedCache;
+
+        private readonly MSALRedisTokenCacheOptions CacheOptions;
+
+        /// <summary>
+        /// Enables the singleton object to access the right HttpContext
+        /// </summary>
+        private IHttpContextAccessor httpContextAccessor;
+
+        /// <summary>Initializes a new instance of the <see cref="MSALPerUserMemoryTokenCache"/> class.</summary>
+        /// <param name="cache">The distributed cache instance</param>
+        public MSALPerUserRedisTokenCacheProvider(IDistributedCache cache, MSALRedisTokenCacheOptions option, IHttpContextAccessor httpContextAccessor)
+        {
+            this.distributedCache = cache;
+            this.httpContextAccessor = httpContextAccessor;
+
+            if (option == null)
+            {
+                throw new NullReferenceException("RedisCacheOptions are not provided");
+            }
+            else
+            {
+                this.CacheOptions = option;
+            }
+        }
+
+        /// <summary>Initializes this instance of TokenCacheProvider with essentials to initialize themselves.</summary>
+        /// <param name="tokenCache">The token cache instance of MSAL application</param>
+        /// <param name="httpcontext">The Httpcontext whose Session will be used for caching.This is required by some providers.</param>
+        /// <param name="user">The signed-in user for whom the cache needs to be established. Not needed by all providers.</param>
+        public void Initialize(ITokenCache tokenCache, HttpContext httpcontext, ClaimsPrincipal user)
+        {
+            tokenCache.SetBeforeAccess(this.UserTokenCacheBeforeAccessNotification);
+            tokenCache.SetAfterAccess(this.UserTokenCacheAfterAccessNotification);
+            tokenCache.SetBeforeWrite(this.UserTokenCacheBeforeWriteNotification);
+        }
+
+        /// <summary>
+        /// Clears the TokenCache's copy of this user's cache.
+        /// </summary>
+        public void Clear(string accountId)
+        {
+            this.distributedCache.Remove(accountId);
```

Comment:
1. missing param validation:
`if (string.IsNullOrWhitespace(accountId)) throw new ArgumentNullException(nameof(accountId), "Cannot be null, empty, or whitespace");

2. will this throw and exception if the accountId is not in the cache?  If so, you should use the TryXXX pattern and check for the presence of the account before removing it.